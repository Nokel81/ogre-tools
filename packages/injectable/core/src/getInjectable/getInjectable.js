import lifecycleEnum from '../dependency-injection-container/lifecycleEnum';

export const injectableSymbol = 'injectable';

export default ({ lifecycle = lifecycleEnum.singleton, ...injectable }) => ({
  aliasType: injectableSymbol,
  lifecycle,
  ...injectable,
});
