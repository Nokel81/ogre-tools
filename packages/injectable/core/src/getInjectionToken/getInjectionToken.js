export const injectionTokenSymbol = 'injection-token';

export default ({ id, decorable = true }) => ({
  id,
  aliasType: injectionTokenSymbol,
  decorable,
});
