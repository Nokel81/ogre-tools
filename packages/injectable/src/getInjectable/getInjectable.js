import identity from 'lodash/fp/identity';

// Note: this function exists only for typed presence in TypeScript.
// It has little purpose in JavaScript.
export default identity;
