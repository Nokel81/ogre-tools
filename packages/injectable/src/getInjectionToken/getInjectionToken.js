// Note: this function exists only for typed presence in TypeScript.
// It has little purpose in JavaScript.
export default ({ id }) => ({ id });
