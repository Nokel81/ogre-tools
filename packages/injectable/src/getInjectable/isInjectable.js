import { injectableSymbol } from './getInjectable';

export default maybeInjectable =>
  maybeInjectable?.aliasType === injectableSymbol;
