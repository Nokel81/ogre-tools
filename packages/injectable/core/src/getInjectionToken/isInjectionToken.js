import { injectionTokenSymbol } from './getInjectionToken';

export default maybeInjectionToken =>
  maybeInjectionToken?.aliasType === injectionTokenSymbol;
