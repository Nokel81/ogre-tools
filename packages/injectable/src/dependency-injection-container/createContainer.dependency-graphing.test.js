import getDi from '../test-utils/getDiForUnitTesting';
import getInjectable from '../getInjectable/getInjectable';
import {
  plantUmlDependencyGraphInjectable,
  registerDependencyGraphing,
} from './extensions/dependency-graphing/dependency-graphing';
import getInjectionToken from '../getInjectionToken/getInjectionToken';

describe('createContainer.dependency-graph', () => {
  it('given dependency graphing, dependencies and injected, creates Plant-UML graph', async () => {
    const parentInjectable = getInjectable({
      id: 'some-parent-injectable',

      instantiate: di => di.inject(childInjectable),
    });

    const childInjectable = getInjectable({
      id: 'some-child-injectable',
      instantiate: di => di.injectMany(injectionToken),
    });

    const injectionToken = getInjectionToken({ id: 'some-injection-token' });
    const tokenInjectable = getInjectable({
      id: 'some-token-injectable',
      instantiate: () => 'irrelevant',
      injectionToken,
    });

    const setuppableInjectable = getInjectable({
      id: 'some-setuppable',
      instantiate: () => 'irrelevant',
      setup: di => {
        di.inject(childInjectable);
        di.inject(setuppableInjectable);
      },
    });

    const di = getDi(
      parentInjectable,
      childInjectable,
      tokenInjectable,
      setuppableInjectable,
    );

    registerDependencyGraphing(di);

    await di.runSetups();

    di.inject(parentInjectable);

    const graph = di.inject(plantUmlDependencyGraphInjectable);

    expect(graph).toBe(
      [
        '@startuml',
        '"Setup(some-setuppable)" ..up* "some-child-injectable" : Setup',
        '"some-child-injectable" ..up* "some-injection-token" : Setup',
        '"some-injection-token" ..up* "some-token-injectable" : Setup',
        '"Setup(some-setuppable)" ..up* "some-setuppable" : Setup',
        '"some-parent-injectable" --up* "some-child-injectable"',
        '@enduml',
      ].join('\n'),
    );
  });
});
