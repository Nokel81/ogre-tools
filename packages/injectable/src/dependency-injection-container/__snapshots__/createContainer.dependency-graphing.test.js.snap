// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContainer.dependency-graph given dependency graphing, dependencies and injected, creates Plant-UML graph 1`] = `
"@startuml
hide members
hide circle
\\"Setup(some-setuppable)\\" ..up* \\"some-child-injectable\\" : Setup
class \\"Setup(some-setuppable)\\"<Setup>
class \\"some-child-injectable\\"<Transient>
\\"some-child-injectable\\" ..up* \\"some-injection-token\\" : Setup
\\"some-injection-token\\" ..up* \\"some-token-injectable\\" : Setup
class \\"some-injection-token\\"<Token> #orange
class \\"some-token-injectable\\"<Singleton>
\\"Setup(some-setuppable)\\" ..up* \\"some-setuppable\\" : Setup
class \\"some-setuppable\\"<Singleton>
class \\"some-parent-injectable\\"<Singleton>
\\"some-parent-injectable\\" --up* \\"some-child-injectable\\"
\\"some-child-injectable\\" --up* \\"some-injection-token\\"
\\"some-injection-token\\" --up* \\"some-token-injectable\\"
@enduml"
`;
